### PR TITLE
Ensure that all handlers get the latest state (eventually)

### PR DESCRIPTION
**Pull request by nolar**
Keep track of which state was handled by which handler, and re-execute it if the state is changed during the handling cycle.

> Issue : replays #9, indirectly #161 and #150

## Description

In #9, there was an attempt to make the handler's status persistent. As a side-effect, it also could detect the state changes during the handling cycle. The persistent progress was not needed or not much useful. However, the change detection is needed again. This PR partially replays the same solution.

Currently, if an object is changed during the long handling cycle (which can include multiple handlers and multiple retries — i.e. for hours sometimes), these changes will not be noticed in some handlers. The "last-seen" state will be stored as it was at the moment of the _last_ handler, and only the changes relative to that state will be handled since then. 

**As a result,** if some handler happened to run before the change was introduced, i.e. it has got an old state, that handler would never know about the new state.

With this PR, **all update & deletion handlers** handlers will track the handled state of the objects individually (as a short hash-digest). If it changes during the handling cycle, those handlers will be re-executed with the new state. 

**This guarantees that every handler will see the latest actual state of the object — the closest to the concept of reconciliation.**

**For creation handlers,** the behaviour is slightly different. In case of changes during the creation cycle, the creation handlers will be continued as normally (not re-executed). And only after the full creation cycle is done, an update cycle will start with the changes relatively to the _first_ (not last!) creation handler, i.e. when the object was seen in the first time. 

This should prevent spawning duplicate children objects, which is usually done in the creation handlers.  The update handlers should be designed idempotent anyway, so it is safe to shift the repeated handling of the same state from the creation handlers to the update handlers.

## Types of Changes

- New feature (non-breaking change which adds functionality)

## Review

- [ ] Tests
- [ ] Documentation

### CODE REVIEW

**Comment by nolar** [comment]
Rebased on master with amendments for the new codebase & features.

File: kopf/structs/lastseen.py [hunk 1]
```diff
@@ -98,5 +99,29 @@ def retreive_state(body):
...
+
+def freeze_state(*, body, patch, extra_fields=None):
+    frozen_state = body.get('status', {}).get('kopf', {}).get('frozen-state')
+    actual_state = get_state(body, extra_fields=extra_fields)
+    if frozen_state is None:
+        storage = patch.setdefault('status', {}).setdefault('kopf', {})
+        storage['frozen-state'] = json.dumps(actual_state)
+
+
+def compute_digest(body, extra_fields=None):
```
Comment (psycho-ir) [inline kopf/structs/lastseen.py]:
```suggestion
def compute_digest(body, extra_fields=None) -> str:
```

File: kopf/reactor/causation.py [hunk 2]
```diff
@@ -67,6 +67,7 @@ class Cause(NamedTuple):
     initial: bool
     body: MutableMapping
     patch: MutableMapping
+    digest: Union[str, int]
```
Comment (psycho-ir) [inline kopf/reactor/causation.py]: in what conditions the `digest` can be `int`?

Comment (nolar) [reply kopf/reactor/causation.py]: When it is reimplemented to something else (e.g. crc32). We just show that we do not care that it is necessary a string.

Comment (psycho-ir) [reply kopf/reactor/causation.py]: I see, so in the specification we don't care wether this string or int.

File: kopf/structs/lastseen.py (hunk 1 above)
Comment (nolar) [reply kopf/structs/lastseen.py]: Good idea. I will do this in a separate refactoring PR with all the types put to all the functions all around (not 50/50 like it is now).

Comment (psycho-ir) [reply kopf/structs/lastseen.py]: Fair enough.

**Pull request merged**